### PR TITLE
axis: allow domain prop to be function (RE-1338)

### PR DESCRIPTION
**Pull request by tommilligan**
https://reinfer.atlassian.net/browse/RE-1338

This allow us to render custom domains based on the min/max data values in the chart. We already get these computed in recharts for free, so it seems a shame to waste them by trying to reproduce it ourselves in the calling code.

This PR allows `domain` to be a function, that is given the available data from context in the `recharts` code. We only need access to `dataDomain` now, but `allowDataOverflow` is right there, so let's grab it at the same time.

### CODE REVIEW

File: src/cartesian/XAxis.js [hunk 1]
```diff
@@ -19,12 +19,15 @@ class XAxis extends PureComponent {
     unit: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
     // The unique id of x-axis
     xAxisId: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
-    domain: PropTypes.arrayOf(PropTypes.oneOfType([
-      PropTypes.string,
-      PropTypes.number,
+    domain: PropTypes.oneOfType([
```
Comment (scallaway) [inline src/cartesian/XAxis.js:22]: Cor `PropTypes` - you know it's _old_ React when you're having to define these.

**Review by scallaway — CHANGES_REQUESTED** [review]
I think we need to update the `ChartUtilsSpec.js` file with the new definition of `domain` as well

File: src/cartesian/XAxis.js (hunk 1 above)
Comment (tommilligan) [reply src/cartesian/XAxis.js:22]: IKR, blast from the past

**Comment by tommilligan** [comment]
Updated with a small added unit test.

**Review by scallaway — APPROVED** [review]
Looks good to me! :+1:

**Pull request merged**